opnode/rollup/driver/step.go

```diff
@@ -172,24 +177,99 @@ func (d *outputImpl) step(ctx context.Context, l2Head eth.L2BlockRef, l2Finalize
 			NoTxPool:              false,
 		}
 
-		payload, err := d.addBlock(ctx, fc, attrs, true, updateUnsafeHead)
+		// We are either verifying blocks (with a potential for a reorg) or inserting a safe head to the chain
+		if lastHead.Hash != lastSafeHead.Hash {
+			payload, reorg, err = d.verifySafeBlock(ctx, fc, attrs, lastSafeHead.ID())
+
+		} else {
+			payload, err = d.insertHeadBlock(ctx, fc, attrs, true)
+		}
 		if err != nil {
-			return last, fmt.Errorf("failed to extend L2 chain at block %d/%d of epoch %d: %w", i, len(batches), epoch, err)
+			return lastHead, lastSafeHead, didReorg, fmt.Errorf("failed to extend L2 chain at block %d/%d of epoch %d: %w", i, len(batches), epoch, err)
 		}
+
 		newLast, err := derive.BlockReferences(payload, &d.Config.Genesis)
 		if err != nil {
-			return last, fmt.Errorf("failed to derive block references: %w", err)
+			return lastHead, lastSafeHead, didReorg, fmt.Errorf("failed to derive block references: %w", err)
+		}
+		if reorg {
+			didReorg = true
+		}
+		// If reorg or the L2 Head is not ahead of the safe head, bump the head block.
+		if reorg || lastHead.Hash == lastSafeHead.Hash {
+			lastHead = newLast
+		}
+		lastSafeHead = newLast
+
+		fc.HeadBlockHash = lastHead.Hash
+		fc.SafeBlockHash = lastSafeHead.Hash
+	}
+
+	return lastHead, lastSafeHead, didReorg, nil
+}
+
+// attributesMatchBlock checks if the L2 attributes pre-inputs match the output
+// nil if it is a match. If err is not nil, the error contains the reason for the mismatch
+func attributesMatchBlock(attrs *l2.PayloadAttributes, parentHash common.Hash, block *types.Block) error {
```

Comment:
nit: we can improve the errors by inserting the different values with `%s <> %s` (calls `common.Hash.String()`, and e.g. adding the tx index `i` to know which tx failed to unmarshal.

Reply:
Added `fmt.Errorf` calls here.